examples/js/loaders/GLTFLoader.js

```diff
@@ -538,6 +538,146 @@ THREE.GLTFLoader = ( function () {
 	 *
 	 * Specification: https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_pbrSpecularGlossiness
 	 */
+
+	/**
+	 * A sub class of THREE.StandardMaterial with some of the functionality
+	 * changed via the `onBeforeCompile` callback
+	 * @pailhead
+	 */
+
+	function MeshStandardSGMaterial( params ) {
+
+		THREE.MeshStandardMaterial.call( this );
+
+		this.isGLTFSpecularGlossinessMaterial = true;
+
+		//various chunks that need replacing
+		var specularMapParsFragmentChunk = [
+			'#ifdef USE_SPECULARMAP',
+			'	uniform sampler2D specularMap;',
+			'#endif'
+		].join( '\n' );
+
+		var glossinessMapParsFragmentChunk = [
+			'#ifdef USE_GLOSSINESSMAP',
+			'	uniform sampler2D glossinessMap;',
+			'#endif'
+		].join( '\n' );
+
+		var specularMapFragmentChunk = [
+			'vec3 specularFactor = specular;',
+			'#ifdef USE_SPECULARMAP',
+			'	vec4 texelSpecular = texture2D( specularMap, vUv );',
+			'	texelSpecular = sRGBToLinear( texelSpecular );',
+			'	// reads channel RGB, compatible with a glTF Specular-Glossiness (RGBA) texture',
+			'	specularFactor *= texelSpecular.rgb;',
+			'#endif'
+		].join( '\n' );
+
+		var glossinessMapFragmentChunk = [
+			'float glossinessFactor = glossiness;',
+			'#ifdef USE_GLOSSINESSMAP',
+			'	vec4 texelGlossiness = texture2D( glossinessMap, vUv );',
+			'	// reads channel A, compatible with a glTF Specular-Glossiness (RGBA) texture',
+			'	glossinessFactor *= texelGlossiness.a;',
+			// 'gl_FragColor = vec4(vec3(glossinessFactor),1.);\nreturn;\n',
+			'#endif'
+		].join( '\n' );
+
+		var lightPhysicalFragmentChunk = [
+			'PhysicalMaterial material;',
+			'material.diffuseColor = diffuseColor.rgb;',
+			'material.specularRoughness = clamp( 1.0 - glossinessFactor, 0.04, 1.0 );',
+			'material.specularColor = specularFactor.rgb;',
+		].join( '\n' );
+
+		var uniforms = {
+			specular: { value: new THREE.Color().setHex( 0xffffff ) },
+			glossiness: { value: 1 },
+			specularMap: { value: null },
+			glossinessMap: { value: null }
+		};
+
+		this._extraUniforms = uniforms;
```

Comment:
What's this `._extraUniforms` for?

Reply:
Just a namespace for ones own extra uniforms.

Reply:
Own extra uniforms? For example?

Reply:
No, sorry for the lacking response.

If you look at the webgl modified material example you'll notice a pattern, one that i don't like:
https://github.com/mrdoob/three.js/blob/dev/examples/webgl_materials_modified.html#L83
https://github.com/mrdoob/three.js/blob/dev/examples/webgl_materials_modified.html#L98

There's a global `materialShader` variable where the internal material's shader object is stored. This is either a poor practice or a very limited (incomplete) example. Obviously you cant have the `GLTF.Loader` expose such a variable. `time` uniform is directly attached to that shader object, "manually" - you have to wait for the callback to fire and only then create that uniform object.

I generally use the pattern of:

```
const myExtraMaterialStuff= {}

const myWebGLUniformAbstractions = { foo: { value: 0 } }

myExtraMaterialStuff.extendedUniforms = myWebGLUniformAbstractions

myExtraMaterialStuff.employee = {
  name: 'foo',
  lastname: 'bar',
  id: THREE.Math.generateUUID(),
  title: "Material overseer",
}

const myThreeMaterial = new THREE.SomeMaterial()

myThreeMaterial.userData = myExtraMaterialStuff
```

So if i do end up using `onBeforeCompile` and require additional input, i'd store a `myMaterial.extraInput = { uFoo: {value:0} }` and then just use that directly. You can think of this as:

> I will mixin `THREE.ShaderMaterial` API with some built in material

ie:

```
myShaderMaterial.uniforms.uColor.value.set( getGUIColor() )
//vs
myBuiltInMaterial.color.set( getGUIColor() )
```

The extended material ends up using both these familiar interfaces. `m.color.set()` and `m.uniforms.uTime.value = getGlobalTime()`. I just called it `_extraUniforms`. 

However, in this diff, this is actually redundant, since it can remain private:
https://github.com/pailhead/three.js/blob/902cbb2da38dbb4891e7d87411177096a3f85ee4/examples/js/loaders/GLTFLoader.js#L627

^ accessess `const uniforms` from the closure